File: ts/src/validation.ts
const MaxFieldNameLength = 255;
const MaxOperatorLength = 20;
const MaxValueStringLength = 400000; // DynamoDB item size limit (approx)
const MaxNestedDepth = 32;
const MaxExpressionLength = 4096;

type SecurityValidationErrorType =
  | 'InjectionAttempt'
  | 'InvalidField'
  | 'InvalidOperator'
  | 'InvalidValue'
  | 'InvalidExpression'
  | 'InvalidTableName'
  | 'InvalidIndexName';

export class SecurityValidationError extends Error {
  readonly type: SecurityValidationErrorType;
  readonly detail: string;

  constructor(type: SecurityValidationErrorType, detail: string) {
    super(`security validation failed: ${type}`);
    this.type = type;
    this.detail = detail;
    this.name = 'SecurityValidationError';
  }
}

export function isSecurityValidationError(
  value: unknown,
): value is SecurityValidationError {
  return value instanceof SecurityValidationError;
}

const dangerousPatterns = [
  "'",
  '"',
  ';',
  '--',
  '/*',
  '*/',
  '<script',
  '</script',
  'eval(',
  'expression(',
  'import(',
  'require(',
];

const sqlKeywords = [
  'union',
  'select',
  'insert',
  'update',
  'delete',
  'drop',
  'alter',
  'exec',
  'execute',
  'script',
  'javascript',
  'vbscript',
];

const legitimateFieldPatterns = [
  /^(created|updated)at$/i,
  /^create(d|r)_?(at|time|date)$/i,
  /^update(d|r)_?(at|time|date)$/i,
  /^delete(d|r)_?(at|time|date|flag)$/i,
  /^insert(ed|er)_?(at|time|date)$/i,
  /^select(ed|or)_?(at|time|date)$/i,
];

const valueScriptPatterns = [
  '<script',
  '</script',
  'eval(',
  'expression(',
  'import(',
  'require(',
  'javascript:',
  'vbscript:',
  'onload=',
  'onerror=',
  'onclick=',
];

const valueSQLInjectionPatterns = [
  "'; drop table",
  "'; delete from",
  "'; update ",
  "'; insert into",
  '"; drop table',
  '"; delete from',
  '"; update ',
  '"; insert into',
  "' or 1=1",
  '" or 1=1',
  "' or '1'='1",
  '" or "1"="1',
  '/**/union/**/select',
  'concat(0x',
  'char(',
  'load_file(',
  '--',
];

const allowedOperators = new Set([
  '=',
  '!=',
  '<>',
  '<',
  '<=',
  '>',
  '>=',
  'BETWEEN',
  'IN',
  'BEGINS_WITH',
  'CONTAINS',
  'EXISTS',
  'NOT_EXISTS',
  'ATTRIBUTE_EXISTS',
  'ATTRIBUTE_NOT_EXISTS',
  'EQ',
  'NE',
  'LT',
  'LE',
  'GT',
  'GE',
]);

export {
  MaxExpressionLength,
  MaxFieldNameLength,
  MaxNestedDepth,
  MaxOperatorLength,
  MaxValueStringLength,
};

export function validateFieldName(field: string): void {
  validateFieldNameBasics(field);

  const lower = field.toLowerCase();
  if (containsAnySubstring(lower, dangerousPatterns)) {
    throw new SecurityValidationError(
      'InjectionAttempt',
      'field name contains dangerous pattern',
    );
  }

  validateFieldNameKeywords(lower, field);

  if (containsControlCharacters(field)) {
    throw new SecurityValidationError(
      'InvalidField',
      'field name contains control characters',
    );
  }

  if (field.includes('.')) {
    validateNestedFieldPath(field);
    return;
  }

  validateFieldPart(field);
}

function validateFieldNameBasics(field: string): void {
  if (!field) {
    throw new SecurityValidationError(
      'InvalidField',
      'field name cannot be empty',
    );
  }
  if (field.length > MaxFieldNameLength) {
    throw new SecurityValidationError(
      'InvalidField',
      'field name exceeds maximum length',
    );
  }
}

function validateFieldNameKeywords(fieldLower: string, field: string): void {
  for (const keyword of sqlKeywords) {
    if (!fieldLower.includes(keyword)) continue;
    if (isLegitimateFieldName(field)) continue;
    if (isStandaloneOrSuspiciousKeyword(fieldLower, keyword)) {
      throw new SecurityValidationError(
        'InjectionAttempt',
        'field name contains suspicious content',
      );
    }
  }
}

function isLegitimateFieldName(field: string): boolean {
  return legitimateFieldPatterns.some((p) => p.test(field));
}

function containsControlCharacters(field: string): boolean {
  for (let i = 0; i < field.length; i++) {
    const code = field.charCodeAt(i);
    if ((code >= 0 && code <= 0x1f) || code === 0x7f) return true;
  }
  return false;
}

function validateNestedFieldPath(field: string): void {
  const parts = field.split('.');
  if (parts.length > MaxNestedDepth) {
    throw new SecurityValidationError(
      'InvalidField',
      'nested field depth exceeds maximum',
    );
  }

  for (const part of parts) {
    try {
      validateFieldPart(part);
    } catch {
      throw new SecurityValidationError('InvalidField', 'invalid field part');
    }
  }
}

function isStandaloneOrSuspiciousKeyword(
  fieldLower: string,
  keyword: string,
): boolean {
  if (fieldLower === keyword) return true;

  const suspiciousPatterns = [
    `${keyword};`,
    `;${keyword}`,
    `${keyword} `,
    ` ${keyword}`,
    `${keyword}.`,
    `.${keyword}`,
    `${keyword}-`,
    `-${keyword}`,
  ];

  return suspiciousPatterns.some((p) => fieldLower.includes(p));
}

function validateFieldPart(part: string): void {
  if (!part) {
    throw new SecurityValidationError(
      'InvalidField',
      'field part cannot be empty',
    );
  }

  if (part.includes('[') && part.includes(']')) {
    const open = part.indexOf('[');
    const close = part.lastIndexOf(']');
    if (close <= open) {
      throw new SecurityValidationError(
        'InvalidField',
        'invalid bracket syntax in field part',
      );
    }

    const fieldName = part.slice(0, open);
    const indexPart = part.slice(open + 1, close);
    const remaining = part.slice(close + 1);

    const fieldPattern = /^[a-zA-Z_][a-zA-Z0-9_]*$/;
    if (!fieldPattern.test(fieldName)) {
      throw new SecurityValidationError(
        'InvalidField',
        'field name part must start with letter or underscore and contain only alphanumeric characters and underscores',
      );
    }

    const indexPattern = /^[0-9]+$/;
    if (!indexPattern.test(indexPart)) {
      throw new SecurityValidationError(
        'InvalidField',
        'list index must be a number',
      );
    }

    if (remaining !== '') {
      throw new SecurityValidationError(
        'InvalidField',
        'unexpected characters after list index',
      );
    }

    return;
  }

  const validPattern = /^[a-zA-Z_][a-zA-Z0-9_]*$/;
  if (!validPattern.test(part)) {
    throw new SecurityValidationError(
      'InvalidField',
      'field part must start with letter or underscore and contain only alphanumeric characters and underscores',
    );
  }
}

export function validateOperator(op: string): void {
  if (!op) {
    throw new SecurityValidationError(
      'InvalidOperator',
      'operator cannot be empty',
    );
  }

  if (op.length > MaxOperatorLength) {
    throw new SecurityValidationError(
      'InvalidOperator',
      'operator exceeds maximum length',
    );
  }

  const opUpper = op.trim().toUpperCase();
  if (!allowedOperators.has(opUpper)) {
    throw new SecurityValidationError(
      'InvalidOperator',
      'operator not allowed',
    );
  }

  const opLower = op.toLowerCase();
  if (containsAnySubstring(opLower, dangerousPatterns)) {
    throw new SecurityValidationError(
      'InjectionAttempt',
      'operator contains dangerous pattern',
    );
  }
}

export function validateValue(value: unknown): void {
  if (value === null || value === undefined) return;

  if (typeof value === 'string') {
    validateStringValue(value);
    return;
  }

  if (Array.isArray(value)) {
    validateArrayValue(value);
    return;
  }

  if (typeof value === 'object') {
    validateObjectValue(value as Record<string, unknown>);
    return;
  }

  if (typeof value === 'function' || typeof value === 'symbol') {
    throw new SecurityValidationError('InvalidValue', 'unsupported value type');
  }
}

function validateStringValue(value: string): void {
  if (value.length > MaxValueStringLength) {
    throw new SecurityValidationError(
      'InvalidValue',
      'string value exceeds maximum length',
    );
  }

  const lower = value.toLowerCase();

  if (
    containsAnySubstring(lower, valueScriptPatterns) ||
    (value.includes('/*') && value.includes('*/')) ||
    containsAnySubstring(lower, valueSQLInjectionPatterns) ||
    looksLikeUnionSelectInjection(lower, value)
  ) {
    throw new SecurityValidationError(
      'InjectionAttempt',
      'string value contains dangerous pattern',
    );
  }
}

function validateArrayValue(value: unknown[]): void {
  if (value.length > 100) {
    throw new SecurityValidationError(
      'InvalidValue',
      'slice value exceeds maximum length of 100 items',
    );
  }
  for (const item of value) {
    try {
      validateValue(item);
    } catch {
      throw new SecurityValidationError(
        'InvalidValue',
        'invalid item in collection',
      );
    }
  }
}

function validateObjectValue(value: Record<string, unknown>): void {
  const entries = Object.entries(value);
  if (entries.length > 100) {
    throw new SecurityValidationError(
      'InvalidValue',
      'map value exceeds maximum keys',
    );
  }

  for (const [k, v] of entries) {
    try {
      validateFieldName(k);
    } catch {
      throw new SecurityValidationError('InvalidValue', 'invalid map key');
    }
    try {
      validateValue(v);
    } catch {
      throw new SecurityValidationError('InvalidValue', 'invalid map value');
    }
  }
}

export function validateExpression(expression: string): void {
  if (expression.length > MaxExpressionLength) {
    throw new SecurityValidationError(
      'InvalidExpression',
      'expression exceeds maximum length',
    );
  }

  const lower = expression.toLowerCase();
  if (containsAnySubstring(lower, dangerousPatterns)) {
    throw new SecurityValidationError(
      'InjectionAttempt',
      'expression contains dangerous pattern',
    );
  }

  const sqlInjectionPatterns = [
    'union select',
    'insert into',
    'update set',
    'delete from',
    'drop table',
    'alter table',
    'exec ',
    'execute ',
  ];
  if (containsAnySubstring(lower, sqlInjectionPatterns)) {
    throw new SecurityValidationError(
      'InjectionAttempt',
      'expression contains dangerous pattern',
    );
  }
}

export function validateTableName(name: string): void {
  if (name.length < 3 || name.length > 255) {
    throw new SecurityValidationError(
      'InvalidTableName',
      'table name length invalid',
    );
  }

  const pattern = /^[a-zA-Z0-9_.-]+$/;
  if (!pattern.test(name)) {
    throw new SecurityValidationError(
      'InvalidTableName',
      'table name contains invalid characters',
    );
  }

  const lower = name.toLowerCase();
  if (containsAnySubstring(lower, dangerousPatterns)) {
    throw new SecurityValidationError(
      'InjectionAttempt',
      'table name contains dangerous pattern',
    );
  }
}

export function validateIndexName(name: string): void {
  if (!name) return;

  if (name.length < 3 || name.length > 255) {
    throw new SecurityValidationError(
      'InvalidIndexName',
      'index name length invalid',
    );
  }

  const pattern = /^[a-zA-Z0-9_.-]+$/;
  if (!pattern.test(name)) {
    throw new SecurityValidationError(
      'InvalidIndexName',
      'index name contains invalid characters',
    );
  }
}

function containsAnySubstring(haystack: string, needles: string[]): boolean {
  return needles.some((n) => haystack.includes(n));
}

function looksLikeUnionSelectInjection(lower: string, raw: string): boolean {
  if (!lower.includes('union') || !lower.includes('select')) return false;

  if (
    !lower.includes('union select') &&
    !lower.includes('union all select') &&
    !lower.includes('union/**/select')
  ) {
    return false;
  }

  return (
    lower.includes('from') ||
    lower.includes('*') ||
    raw.endsWith('--') ||
    raw.endsWith(';')
  );
}

File: py/src/tabletheory_py/validation.py
from __future__ import annotations

import re
from collections.abc import Mapping, Sequence
from typing import Any

MaxFieldNameLength = 255
MaxOperatorLength = 20
MaxValueStringLength = 400000  # DynamoDB item size limit (approx)
MaxNestedDepth = 32
MaxExpressionLength = 4096


class SecurityValidationError(Exception):
    def __init__(self, *, type: str, detail: str) -> None:
        super().__init__(f"security validation failed: {type}")
        self.type = type
        self.detail = detail


_DANGEROUS_PATTERNS = (
    "'",
    '"',
    ";",
    "--",
    "/*",
    "*/",
    "<script",
    "</script",
    "eval(",
    "expression(",
    "import(",
    "require(",
)

_SQL_KEYWORDS = (
    "union",
    "select",
    "insert",
    "update",
    "delete",
    "drop",
    "alter",
    "exec",
    "execute",
    "script",
    "javascript",
    "vbscript",
)

_LEGITIMATE_FIELD_PATTERNS = (
    re.compile(r"^(created|updated)at$", re.IGNORECASE),
    re.compile(r"^create(d|r)_?(at|time|date)$", re.IGNORECASE),
    re.compile(r"^update(d|r)_?(at|time|date)$", re.IGNORECASE),
    re.compile(r"^delete(d|r)_?(at|time|date|flag)$", re.IGNORECASE),
    re.compile(r"^insert(ed|er)_?(at|time|date)$", re.IGNORECASE),
    re.compile(r"^select(ed|or)_?(at|time|date)$", re.IGNORECASE),
)

_VALUE_SCRIPT_PATTERNS = (
    "<script",
    "</script",
    "eval(",
    "expression(",
    "import(",
    "require(",
    "javascript:",
    "vbscript:",
    "onload=",
    "onerror=",
    "onclick=",
)

_VALUE_SQL_INJECTION_PATTERNS = (
    "'; drop table",
    "'; delete from",
    "'; update ",
    "'; insert into",
    '"; drop table',
    '"; delete from',
    '"; update ',
    '"; insert into',
    "' or 1=1",
    '" or 1=1',
    "' or '1'='1",
    '" or "1"="1',
    "/**/union/**/select",
    "concat(0x",
    "char(",
    "load_file(",
    "--",
)

_ALLOWED_OPERATORS = {
    "=",
    "!=",
    "<>",
    "<",
    "<=",
    ">",
    ">=",
    "BETWEEN",
    "IN",
    "BEGINS_WITH",
    "CONTAINS",
    "EXISTS",
    "NOT_EXISTS",
    "ATTRIBUTE_EXISTS",
    "ATTRIBUTE_NOT_EXISTS",
    "EQ",
    "NE",
    "LT",
    "LE",
    "GT",
    "GE",
}


def validate_field_name(field: str) -> None:
    _validate_field_name_basics(field)

    lower = field.lower()
    if _contains_any_substring(lower, _DANGEROUS_PATTERNS):
        raise SecurityValidationError(type="InjectionAttempt", detail="field name contains dangerous pattern")

    _validate_field_name_keywords(lower, field)

    if _contains_control_characters(field):
        raise SecurityValidationError(type="InvalidField", detail="field name contains control characters")

    if "." in field:
        _validate_nested_field_path(field)
        return

    _validate_field_part(field)


def _validate_field_name_basics(field: str) -> None:
    if not field:
        raise SecurityValidationError(type="InvalidField", detail="field name cannot be empty")
    if len(field) > MaxFieldNameLength:
        raise SecurityValidationError(type="InvalidField", detail="field name exceeds maximum length")


def _validate_field_name_keywords(field_lower: str, field: str) -> None:
    for kw in _SQL_KEYWORDS:
        if kw not in field_lower:
            continue
        if _is_legitimate_field_name(field):
            continue
        if _is_standalone_or_suspicious_keyword(field_lower, kw):
            raise SecurityValidationError(
                type="InjectionAttempt", detail="field name contains suspicious content"
            )


def _is_legitimate_field_name(field: str) -> bool:
    return any(p.match(field) is not None for p in _LEGITIMATE_FIELD_PATTERNS)


def _contains_control_characters(value: str) -> bool:
    for ch in value:
        code = ord(ch)
        if 0 <= code <= 0x1F or code == 0x7F:
            return True
    return False


def _validate_nested_field_path(field: str) -> None:
    parts = field.split(".")
    if len(parts) > MaxNestedDepth:
        raise SecurityValidationError(type="InvalidField", detail="nested field depth exceeds maximum")

    for part in parts:
        try:
            _validate_field_part(part)
        except SecurityValidationError as err:
            raise SecurityValidationError(type="InvalidField", detail="invalid field part") from err


def _is_standalone_or_suspicious_keyword(field_lower: str, keyword: str) -> bool:
    if field_lower == keyword:
        return True

    suspicious = (
        f"{keyword};",
        f";{keyword}",
        f"{keyword} ",
        f" {keyword}",
        f"{keyword}.",
        f".{keyword}",
        f"{keyword}-",
        f"-{keyword}",
    )
    return _contains_any_substring(field_lower, suspicious)


def _validate_field_part(part: str) -> None:
    if not part:
        raise SecurityValidationError(type="InvalidField", detail="field part cannot be empty")

    if "[" in part and "]" in part:
        open_bracket = part.find("[")
        close_bracket = part.rfind("]")
        if close_bracket <= open_bracket:
            raise SecurityValidationError(type="InvalidField", detail="invalid bracket syntax in field part")

        field_name = part[:open_bracket]
        index_part = part[open_bracket + 1 : close_bracket]
        remaining = part[close_bracket + 1 :]

        if re.match(r"^[a-zA-Z_][a-zA-Z0-9_]*$", field_name) is None:
            raise SecurityValidationError(
                type="InvalidField",
                detail=(
                    "field name part must start with letter or underscore and contain only alphanumeric characters "
                    "and underscores"
                ),
            )

        if re.match(r"^[0-9]+$", index_part) is None:
            raise SecurityValidationError(type="InvalidField", detail="list index must be a number")

        if remaining:
            raise SecurityValidationError(
                type="InvalidField", detail="unexpected characters after list index"
            )
        return

    if re.match(r"^[a-zA-Z_][a-zA-Z0-9_]*$", part) is None:
        raise SecurityValidationError(
            type="InvalidField",
            detail="field part must start with letter or underscore and contain only alphanumeric characters and underscores",
        )


def validate_operator(op: str) -> None:
    if not op:
        raise SecurityValidationError(type="InvalidOperator", detail="operator cannot be empty")
    if len(op) > MaxOperatorLength:
        raise SecurityValidationError(type="InvalidOperator", detail="operator exceeds maximum length")

    op_upper = op.strip().upper()
    if op_upper not in _ALLOWED_OPERATORS:
        raise SecurityValidationError(type="InvalidOperator", detail="operator not allowed")

    op_lower = op.lower()
    if _contains_any_substring(op_lower, _DANGEROUS_PATTERNS):
        raise SecurityValidationError(type="InjectionAttempt", detail="operator contains dangerous pattern")


def validate_value(value: Any) -> None:
    if value is None:
        return

    if isinstance(value, str):
        _validate_string_value(value)
        return

    if isinstance(value, (bytes, bytearray, memoryview)):
        return

    if isinstance(value, Sequence) and not isinstance(value, (str, bytes, bytearray, dict)):
        _validate_sequence_value(value)
        return

    if isinstance(value, Mapping):
        _validate_mapping_value(value)
        return

    if callable(value) or isinstance(value, complex):
        raise SecurityValidationError(type="InvalidValue", detail="unsupported value type")


def _validate_string_value(value: str) -> None:
    if len(value) > MaxValueStringLength:
        raise SecurityValidationError(type="InvalidValue", detail="string value exceeds maximum length")

    lower = value.lower()

    if (
        _contains_any_substring(lower, _VALUE_SCRIPT_PATTERNS)
        or ("/*" in value and "*/" in value)
        or _contains_any_substring(lower, _VALUE_SQL_INJECTION_PATTERNS)
        or _looks_like_union_select_injection(lower, value)
    ):
        raise SecurityValidationError(
            type="InjectionAttempt", detail="string value contains dangerous pattern"
        )


def _validate_sequence_value(value: Sequence[Any]) -> None:
    if len(value) > 100:
        raise SecurityValidationError(
            type="InvalidValue", detail="slice value exceeds maximum length of 100 items"
        )

    for item in value:
        try:
            validate_value(item)
        except SecurityValidationError as err:
            raise SecurityValidationError(type="InvalidValue", detail="invalid item in collection") from err


def _validate_mapping_value(value: Mapping[Any, Any]) -> None:
    if len(value) > 100:
        raise SecurityValidationError(type="InvalidValue", detail="map value exceeds maximum keys")

    for k, v in value.items():
        try:
            validate_field_name(str(k))
        except SecurityValidationError as err:
            raise SecurityValidationError(type="InvalidValue", detail="invalid map key") from err
        try:
            validate_value(v)
        except SecurityValidationError as err:
            raise SecurityValidationError(type="InvalidValue", detail="invalid map value") from err


def validate_expression(expression: str) -> None:
    if len(expression) > MaxExpressionLength:
        raise SecurityValidationError(type="InvalidExpression", detail="expression exceeds maximum length")

    lower = expression.lower()
    if _contains_any_substring(lower, _DANGEROUS_PATTERNS):
        raise SecurityValidationError(type="InjectionAttempt", detail="expression contains dangerous pattern")

    sql_injection_patterns = (
        "union select",
        "insert into",
        "update set",
        "delete from",
        "drop table",
        "alter table",
        "exec ",
        "execute ",
    )
    if _contains_any_substring(lower, sql_injection_patterns):
        raise SecurityValidationError(type="InjectionAttempt", detail="expression contains dangerous pattern")


def validate_table_name(name: str) -> None:
    if len(name) < 3 or len(name) > 255:
        raise SecurityValidationError(type="InvalidTableName", detail="table name length invalid")

    if re.match(r"^[a-zA-Z0-9_.-]+$", name) is None:
        raise SecurityValidationError(
            type="InvalidTableName", detail="table name contains invalid characters"
        )

    lower = name.lower()
    if _contains_any_substring(lower, _DANGEROUS_PATTERNS):
        raise SecurityValidationError(type="InjectionAttempt", detail="table name contains dangerous pattern")


def validate_index_name(name: str) -> None:
    if not name:
        return

    if len(name) < 3 or len(name) > 255:
        raise SecurityValidationError(type="InvalidIndexName", detail="index name length invalid")

    if re.match(r"^[a-zA-Z0-9_.-]+$", name) is None:
        raise SecurityValidationError(
            type="InvalidIndexName", detail="index name contains invalid characters"
        )


def _contains_any_substring(haystack: str, needles: Sequence[str]) -> bool:
    return any(n in haystack for n in needles)


def _looks_like_union_select_injection(lower: str, raw: str) -> bool:
    if "union" not in lower or "select" not in lower:
        return False

    if "union select" not in lower and "union all select" not in lower and "union/**/select" not in lower:
        return False

    return "from" in lower or "*" in lower or raw.endswith("--") or raw.endswith(";")
